### ISSUE TITLE
Confusing error when trying to re-borrow a function argument

### ISSUE THREAD

**Issue by JaniM** [rust-lang/rust#75871]
It is a common mistake, especially for beginners, to attempt to reborrow function arguments when calling other functions. 

I tried this code:

```rust
struct T;
fn foo(x: &mut T) {}
fn bar(x: &mut T) {
    foo(&mut x)
}
```

I expected to see this happen: Compile error following what the above code does if `foo(&x)` is attempted instead.

```
error[Exxxx]: mismatched types
 --> src/lib.rs:6:9
  |
6 |     foo(&mut x)
  |         ^^ expected `&mut T`, found `&mut &mut T`
  |
  = note; you don't need to use `&mut` when you already have a mutable reference
```

Instead, this happened: Misleading error.

```
error[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
 --> src/lib.rs:6:9
  |
5 | fn bar(x: &mut T) {
  |        - help: consider changing this to be mutable: `mut x`
6 |     foo(&mut x)
  |         ^^^^^^ cannot borrow as mutable
```

### Meta
1.45.2 srable & 1.47.0 nightly on the playground

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=768e5a2d690393f89aed38be3fa5dee6

**Comment by pickfire** [comment]
Seemed similar to https://github.com/rust-lang/rust/issues/71389 except that that is `*x = xxx` but this is `f(x)`.

**Comment by estebank** [comment]
It not _incorrect_, just stylistically suspect given the circumstances. `&mut x` will reborrow, and [it is what to use if you were calling `foo` multiple times](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=749e2342f0ead7af6d8901cf0de19806). Only in this case, where you only need to pass `x` around once is the suggestion not what you want to do.
